server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -242,12 +258,16 @@ public RootObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperMergeC
             dynamicTemplates = this.dynamicTemplates;
         }
         final Map<String, RuntimeField> runtimeFields = new HashMap<>(this.runtimeFields);
-        assert this.runtimeFields != mergeWithObject.runtimeFields;
+        assert this.runtimeFields != mergeWithObject.runtimeFields || this.runtimeFields == Map.<String, RuntimeField>of();
```

Comment:
Both `this.runtimeFields` and `mergeWithObject.runtimeFields` can have the same reference - to `Map.of()` (immutable empty map). In this case we shouldn't throw an assertion error.

This happens in the context of `org.elasticsearch.index.mapper.Mapping#withFieldsBudget`. We're calling `org.elasticsearch.index.mapper.RootObjectMapper#withoutMappers` which provides `Map.of()` for the `runtimeFields` argument. Even if we changed that and provided a `new HashMap<>()`, we'd run into a similar situation because in `org.elasticsearch.index.mapper.RootObjectMapper#merge`, we're doing `Map.copyOf(runtimeFields)`, which returns the equivalent of `Map.of()` in case you provide an empty map.

Maybe we should just drop the assertion entirely?

Reply:
Thanks for the context. The assert was added in #65378 , back when we were still cloning objects mappers. Clone has been removed with #101485 , hence I think the assert can be removed too.

Reply:
done